### DISCUSSION TITLE
The problem of the touch screen function of ILI9341 in the LVGL mode

### DISCUSSION THREAD

**Discussion by davidgxj** [arendst/Tasmota#17344] [General]
LVGL is a good feature. I use ILI9341 touch screen, including XPT2046.
It may not be able to test the touch function.

{"NAME":"Lolin TFT 2.4 Touch Shield","GPIO":[224,225,226,1,1,800,1,1,1,1,1,1,1,7264,736,672,0,1,992,704,0,1,768,1,0,0,0,0,1,1,1,1,1,0,6210,1],"FLAG":0,"BASE":1}

ili9341 2.4in  240*320

clk=gpio18
mosi=gpio23
res=en
dc=gpio15
blk=gpio22
miso=gpio19
cs1=gpio26 (ili9341 cs)
cs2=gpio17(xpt2046 cs)

fireware LVGL
display.ini=ILI9341_display.ini

I use the touch screen to calibrate DisplayCalibrate.tapp unsuccessful
https://tasmota.github.io/docs/lvgl/#tasmota-port

Tip: https://tasmota.github.io/docs/berry-cookbook/#full-Example "LVGL TouchScreen with 3 Relays" code has bug

**Comment by sfromis** [comment]
If you are not getting any touch action, you may not have wired the screen to have touch connected to the SPI bus, MOSI, MISO, CLK in addition to CS?

If you think that https://tasmota.github.io/docs/Berry-Cookbook/#full-example (correct case is important for links) has a bug, what bug do you see? Anyway the docs pages are open to edits if you see something to improve.

**Reply by davidgxj** [reply]
ILI9341 All SPI bus, MOSI, MISO, CLK, and CS are connected to normal, and the screen using the LVGL page can be displayed normally. But XPT2046 cannot test the touch screen function.

![image](https://user-images.githubusercontent.com/17197318/206895640-d53a5c7c-ea38-4401-a12b-00b0a50f44f6.png)

**Reply by sfromis** [reply]
Not clear if I managed to get across my suggestion of connecting MOSI, MISO and CLK from the ESP32 to **both** the display SPI pins **and** the touch SPI pins? In addition to each of them having a separate CS, of course.

**Reply by sfromis** [reply]
About the `lv_color` not defined, I think that it needs to be `lv.color` instead of `lv_color`; at least `lv.color` exists and seems to do something relevant, returning an instance of the `lv_color` class. Try changing that....

**Reply by davidgxj** [reply]
The ILI9341 I use is similar to the hardware of Lolin TFT 2.4 Touch Shield Display. The others are normal connections, but Xpt2046 has tried various ways in Tasmota Doc, and has not succeeded.

I tried "LVGL TouchScreen with 3 Relays"Code fixation.
![image](https://user-images.githubusercontent.com/17197318/206897983-10be3b20-2167-4ee9-92c7-a2d636922993.png)

**Reply by sfromis** [reply]
The CS pin from touch should be `XPT2046 CS`. Are there any other ways from Tasmota docs?

What is your **exact** wiring?
You did not answer to my suggestion of connecting the SPI from ESP32 to both display **and** touch. Already having those connections is not "to be expected" on ILI9341 boards, as there are separate pins to connect to.

**Reply by sfromis** [reply]
Looks like the cookbook example of [LVGL Touchscreen with 3 Relays](https://tasmota.github.io/docs/Berry-Cookbook/#lvgl-touchscreen-with-3-relays) needs several updates to match current LVGL, @s-hadinger

**Reply by davidgxj** [reply]
I refer to Lolin TFT 2.4 Touch Shield Display
https://templates.blakadder.com/lolin_tft_2_4.html

It's just that it uses the Display mode, the LVGA mode I use.

The pins of TFT 2.4 touch display are defined as follows:
GND
VCC
CLK = LCD_CLK
MOSI = LCD_MOSI
Res = LCD_RESET
Dc = LCD_DC
BLK = LCD_LED
Miso = Touch data read in
CS1 = ILI9341_CS
CS2 = xpt2046_CS
Pen = touch interrupt signal

![image](https://user-images.githubusercontent.com/17197318/206909149-67d9dc60-ea4b-4551-8fc7-63331f9387e5.png)

**My LVGL effect:**
![image](https://user-images.githubusercontent.com/17197318/206909283-eaf40e29-9ef2-4c64-90d1-9e8807cee17c.png)
![image](https://user-images.githubusercontent.com/17197318/206909291-544314eb-4d33-45fc-8d6e-7c903eb40926.png)
![image](https://user-images.githubusercontent.com/17197318/206909296-d282ded8-99df-42c8-bb0c-13ddf146947e.png)
![image](https://user-images.githubusercontent.com/17197318/206909309-64d0a49d-46b7-4df8-a3ae-2237c33fe888.png)
![image](https://user-images.githubusercontent.com/17197318/206909366-68798ae5-2672-4365-931d-95f25451a307.png)

display.ini
`:H,ILI9341,240,320,16,SPI,1,*,*,*,*,*,*,*,40
:S,2,1,1,0,40,20
:I
EF,3,03,80,02
CF,3,00,C1,30
ED,4,64,03,12,81
E8,3,85,00,78
CB,5,39,2C,00,34,02
F7,1,20
EA,2,00,00
C0,1,23
C1,1,10
C5,2,3e,28
C7,1,86
36,1,48
37,1,00
3A,1,55
B1,2,00,18
B6,3,08,82,27
F2,1,00
26,1,01
E0,0F,0F,31,2B,0C,0E,08,4E,F1,37,07,10,03,0E,09,00
E1,0F,00,0E,14,03,11,07,31,C1,48,08,0F,0C,31,36,0F
11,80
29,80
:o,28
:O,29
:A,2A,2B,2C,16
:R,36
:0,48,00,00,00
:1,28,00,00,01
:2,88,00,00,02
:3,E8,00,00,03
:i,20,21
:M
#
`
